login-service/src/backend-services/ims-user-finding.service.ts

```diff
@@ -0,0 +1,348 @@
+import { Injectable } from "@nestjs/common";
+import { GraphqlService } from "src/model/graphql/graphql.service";
+import { StrategyInstance } from "src/model/postgres/StrategyInstance.entity";
+import { UserLoginData } from "src/model/postgres/UserLoginData.entity";
+import { UserLoginDataImsUser } from "src/model/postgres/UserLoginDataImsUser.entity";
+import { UserLoginDataImsUserService } from "src/model/services/user-login-data-ims-user";
+import { UserLoginDataService } from "src/model/services/user-login-data.service";
+import { StrategiesService } from "src/model/services/strategies.service";
+import { Strategy } from "src/strategies/Strategy";
+import { jsonFieldArrayToObject, objectToJsonFieldArray } from "./JSONField";
+import { BackendUserService } from "./backend-user.service";
+
+@Injectable()
+export class ImsUserFindingService {
+    constructor(
+        private readonly graphqlService: GraphqlService,
+        private readonly strategiesService: StrategiesService,
+        private readonly loginDataService: UserLoginDataService,
+        private readonly imsUserService: UserLoginDataImsUserService,
+        private readonly backendUserService: BackendUserService,
+    ) {}
+
+    /**
+     * Checks if the values of the specified fields on the templatedValues match the values of those fields on the node
+     * Deletes the keys that were compared from `templatedValue`
+     *
+     * Can be used directly on e.g. the ims as node to chech the description or on the templated fields of the ims to check those
+     *
+     * @param node The node object containing all values of keys to potentially check
+     * @param requiredTemplatedValues The values that the node object must match to pass.
+     * All fields to check will be deleted from this object
+     * @param fieldsToCheck The names of the fields that should be checked for equivalence on the node object directly instead of on its templated fields.
+     * If not given, all keys of `templatedFields` will be matched
+     * @returns `true` iff all specified keys that the templatedValues had a value fore matched, `false` else
+     */
+    private checkFieldsDirectly(node: object, requiredTemplatedValues: object, fieldsToCheck?: string[]): boolean {
+        const fields = fieldsToCheck ? fieldsToCheck : Object.keys(requiredTemplatedValues);
+        for (const key of fields) {
+            if (requiredTemplatedValues[key]) {
+                if (!node[key]) {
+                    return false;
+                }
+                if (requiredTemplatedValues[key] != node[key]) {
+                    return false;
+                }
+                if (JSON.stringify(node[key]) != JSON.stringify(requiredTemplatedValues[key])) {
```

Comment:
Then this algorithm would be even more computationally intensve than it already is

If you have an idea to make it better in another place, I'm happy to hear about it. Then we might have the performance to sort keys

Reply:
I also sumbeled across that post.
But guess what lowdash does internally

Correct 🎉 , they iterate over every key perform extensive equivalence checks and if it is not equivalend but an object, they recurse down and repeat.
See https://github.com/lodash/lodash/blob/master/.internal/equalObjects.js#L63 and https://github.com/lodash/lodash/blob/master/.internal/baseIsEqual.js#L25

I don't think that would exactly _improve_ performane. But if you insist, I'll use it

Reply:
I don't think that it is that expensive  
The recursing down is also necessary in the current implementation due to the stringify.
The overhead of that comparison would be, that there is additional logarithmic performance impact due to the hashmap lookup foreach entry.
So basically have have an O(n log (n)) instead of O(n).
However, log(x) is typically so small that it does not matter. Especially if you consider that e.g. a string comparison of a large number is much more expensive that the comparison of the number itself.
And in case it does not match, a native implementation can return false immediately, while the stringify version always has to stringify the whole object, and can only return early when comparing the string itself (which is likely cheaper).

Reply:
Ok, I'll change it.

Reply:
I however don't really like, that they had their last npm deploy 2 years ago

https://www.npmjs.com/package/lodash

Reply:
Changed to https://www.npmjs.com/package/fast-equals in 285913b